<a href="https://colab.research.google.com/github/sidh26/DL-Assignments/blob/master/Expt%206a%20-%20Keras_Activations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
activations={}

#Usage of activations

Activations can either be used through an Activation layer, or through the activation argument supported by all forward layers:



```
from keras.layers import Activation, Dense
model.add(Dense(64))
model.add(Activation('tanh'))
```



This is equivalent to:

```
model.add(Dense(64, activation='tanh'))
```

You can also pass an element-wise TensorFlow/Theano/CNTK function as an activation:

```
from keras import backend as K
model.add(Dense(64, activation=K.tanh))
```

#elu

keras.activations.elu(x, alpha=1.0)

Exponential linear unit.

##Arguments

1. x: Input tensor.
2. alpha: A scalar, slope of negative section.

Returns

1. The exponential linear activation: x if x > 0 and alpha * (exp(x)-1) if x < 0.

In [17]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['elu'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.2892 - acc: 0.9125 - val_loss: 0.1470 - val_acc: 0.9546
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1206 - acc: 0.9633 - val_loss: 0.1175 - val_acc: 0.9642
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0815 - acc: 0.9742 - val_loss: 0.0826 - val_acc: 0.9751
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0629 - acc: 0.9800 - val_loss: 0.0818 - val_acc: 0.9738
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0491 - acc: 0.9840 - val_loss: 0.0869 - val_acc: 0.9745
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0404 - acc: 0.9869 - val_loss: 0.0837 - val_acc: 0.9758
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0328 - acc: 0.9895 - val_loss: 0.0727 - val_acc

#softmax

keras.activations.softmax(x, axis=-1)

##Softmax activation function.

##Arguments

1. x: Input tensor.
2. axis: Integer, axis along which the softmax normalization is applied.

##Returns

Tensor, output of softmax transformation.

##Raises

ValueError: In case dim(x) == 1.

In [18]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['softmax'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 2.2690 - acc: 0.1781 - val_loss: 2.1950 - val_acc: 0.6540
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 2.0404 - acc: 0.6707 - val_loss: 1.8526 - val_acc: 0.6757
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 1.6474 - acc: 0.6759 - val_loss: 1.4382 - val_acc: 0.6779
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 1.2600 - acc: 0.6794 - val_loss: 1.0929 - val_acc: 0.6833
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.9764 - acc: 0.6816 - val_loss: 0.8748 - val_acc: 0.6850
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.8105 - acc: 0.6892 - val_loss: 0.7612 - val_acc: 0.6907
Epoch 7/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.7287 - acc: 0.6965 - val_loss: 0.7084 - val_acc

#selu

keras.activations.selu(x)

Scaled Exponential Linear Unit (SELU).

SELU is equal to: scale * elu(x, alpha), where alpha and scale are predefined constants. The values of alpha and scale are chosen so that the mean and variance of the inputs are preserved between two consecutive layers as long as the weights are initialized correctly (see lecun_normal initialization) and the number of inputs is "large enough" (see references for more information).

##Arguments

1. x: A tensor or variable to compute the activation function for.

##Returns

The scaled exponential unit activation: scale * elu(x, alpha).

##Note

1. To be used together with the initialization "lecun_normal".
2. To be used together with the dropout variant "AlphaDropout".

In [19]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['selu'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.3331 - acc: 0.9002 - val_loss: 0.1676 - val_acc: 0.9463
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1370 - acc: 0.9583 - val_loss: 0.1141 - val_acc: 0.9651
Epoch 3/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0937 - acc: 0.9714 - val_loss: 0.0918 - val_acc: 0.9709
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0735 - acc: 0.9766 - val_loss: 0.0854 - val_acc: 0.9744
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0601 - acc: 0.9810 - val_loss: 0.0866 - val_acc: 0.9735
Epoch 6/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0487 - acc: 0.9840 - val_loss: 0.0968 - val_acc: 0.9727
Epoch 7/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.0410 - acc: 0.9862 - val_loss: 0.0958 - val_acc

#softplus

keras.activations.softplus(x)

Softplus activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softplus activation: log(exp(x) + 1).

In [20]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['softplus'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.4559 - acc: 0.8604 - val_loss: 0.2184 - val_acc: 0.9332
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1742 - acc: 0.9465 - val_loss: 0.1506 - val_acc: 0.9487
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1132 - acc: 0.9647 - val_loss: 0.0956 - val_acc: 0.9695
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0833 - acc: 0.9741 - val_loss: 0.1012 - val_acc: 0.9690
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0660 - acc: 0.9794 - val_loss: 0.0881 - val_acc: 0.9716
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0530 - acc: 0.9835 - val_loss: 0.0826 - val_acc: 0.9741
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0435 - acc: 0.9864 - val_loss: 0.0707 - val_acc

#softsign

keras.activations.softsign(x)

##Softsign activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softsign activation: x / (abs(x) + 1).

In [21]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['softsign'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2946 - acc: 0.9129 - val_loss: 0.1648 - val_acc: 0.9509
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1448 - acc: 0.9560 - val_loss: 0.1166 - val_acc: 0.9653
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0995 - acc: 0.9696 - val_loss: 0.1050 - val_acc: 0.9686
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0735 - acc: 0.9776 - val_loss: 0.0815 - val_acc: 0.9750
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0564 - acc: 0.9826 - val_loss: 0.0712 - val_acc: 0.9779
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0439 - acc: 0.9867 - val_loss: 0.0685 - val_acc: 0.9785
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0349 - acc: 0.9897 - val_loss: 0.0697 - val_acc

#relu

keras.activations.relu(x, alpha=0.0, max_value=None, threshold=0.0)

##Rectified Linear Unit.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = alpha * (x - threshold) otherwise.

##Arguments

1. x: Input tensor.
2. alpha: float. Slope of the negative part. Defaults to zero.
3. max_value: float. Saturation threshold.
4. threshold: float. Threshold value for thresholded activation.

##Returns

A tensor.

In [22]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['relu'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.2245 - acc: 0.9303 - val_loss: 0.1006 - val_acc: 0.9699
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0840 - acc: 0.9746 - val_loss: 0.0778 - val_acc: 0.9767
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0544 - acc: 0.9834 - val_loss: 0.0694 - val_acc: 0.9801
Epoch 4/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0405 - acc: 0.9871 - val_loss: 0.0749 - val_acc: 0.9800
Epoch 5/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0294 - acc: 0.9909 - val_loss: 0.0691 - val_acc: 0.9809
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0228 - acc: 0.9925 - val_loss: 0.0835 - val_acc: 0.9786
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0177 - acc: 0.9945 - val_loss: 0.0880 - val_acc

#tanh

keras.activations.tanh(x)

##Hyperbolic tangent activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The hyperbolic activation: tanh(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x))

In [23]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['tanh'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.2949 - acc: 0.9107 - val_loss: 0.1484 - val_acc: 0.9560
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1251 - acc: 0.9624 - val_loss: 0.1087 - val_acc: 0.9663
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0818 - acc: 0.9752 - val_loss: 0.0866 - val_acc: 0.9726
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0589 - acc: 0.9817 - val_loss: 0.0813 - val_acc: 0.9738
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0433 - acc: 0.9865 - val_loss: 0.0701 - val_acc: 0.9763
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0327 - acc: 0.9904 - val_loss: 0.0653 - val_acc: 0.9779
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0237 - acc: 0.9930 - val_loss: 0.0652 - val_acc

#sigmoid

keras.activations.sigmoid(x)

##Sigmoid activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The sigmoid activation: 1 / (1 + exp(-x)).

In [24]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['sigmoid'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.5008 - acc: 0.8583 - val_loss: 0.2657 - val_acc: 0.9171
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2191 - acc: 0.9341 - val_loss: 0.1769 - val_acc: 0.9455
Epoch 3/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.1584 - acc: 0.9520 - val_loss: 0.1337 - val_acc: 0.9586
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1201 - acc: 0.9635 - val_loss: 0.1185 - val_acc: 0.9642
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0946 - acc: 0.9719 - val_loss: 0.0952 - val_acc: 0.9712
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0763 - acc: 0.9773 - val_loss: 0.0846 - val_acc: 0.9734
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0637 - acc: 0.9804 - val_loss: 0.0845 - val_acc

#hard_sigmoid

keras.activations.hard_sigmoid(x)

##Hard sigmoid activation function.

Faster to compute than sigmoid activation.

##Arguments

1. x: Input tensor.

##Returns

1. Hard sigmoid activation:

```
0 if x < -2.5
1 if x > 2.5
0.2 * x + 0.5 if -2.5 <= x <= 2.5.
```

In [25]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['hard_sigmoid'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.5027 - acc: 0.8548 - val_loss: 0.2447 - val_acc: 0.9266
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.2120 - acc: 0.9368 - val_loss: 0.1697 - val_acc: 0.9482
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1481 - acc: 0.9557 - val_loss: 0.1386 - val_acc: 0.9587
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1105 - acc: 0.9667 - val_loss: 0.1037 - val_acc: 0.9684
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0859 - acc: 0.9739 - val_loss: 0.0901 - val_acc: 0.9718
Epoch 6/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.0689 - acc: 0.9791 - val_loss: 0.0798 - val_acc: 0.9756
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0565 - acc: 0.9828 - val_loss: 0.0762 - val_acc

#exponential

keras.activations.exponential(x)

Exponential (base e) activation function.

##Arguments

1. x: Input tensor.

##Returns

1. Exponential activation: exp(x).

In [26]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['exponential'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 14.4537 - acc: 0.1023 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.4902 - val_acc: 0.1010
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.4711 - acc: 0.1022 - val_loss: 14.

#linear

keras.activations.linear(x)

Linear (i.e. identity) activation function.

##Arguments

1. x: Input tensor.

##Returns

Input tensor, unchanged.

In [27]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['linear'] = score[1]

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.4283 - acc: 0.8758 - val_loss: 0.3924 - val_acc: 0.8905
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.3447 - acc: 0.9028 - val_loss: 0.3183 - val_acc: 0.9113
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3217 - acc: 0.9091 - val_loss: 0.3717 - val_acc: 0.8963
Epoch 4/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.3080 - acc: 0.9133 - val_loss: 0.2937 - val_acc: 0.9198
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.2994 - acc: 0.9153 - val_loss: 0.3169 - val_acc: 0.9073
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.2941 - acc: 0.9177 - val_loss: 0.3071 - val_acc: 0.9131
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.2884 - acc: 0.9188 - val_loss: 0.2879 - val_acc

#LeakyReLU

keras.layers.LeakyReLU(alpha=0.3)

Leaky version of a Rectified Linear Unit.

It allows a small gradient when the unit is not active: 
```
f(x) = alpha * x for x < 0
f(x) = x for x >= 0.
```

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

alpha: float >= 0. Negative slope coefficient.

In [28]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['LeakyReLU'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2637 - acc: 0.9205 - val_loss: 0.1555 - val_acc: 0.9533
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1260 - acc: 0.9609 - val_loss: 0.1312 - val_acc: 0.9593
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0942 - acc: 0.9712 - val_loss: 0.0977 - val_acc: 0.9708
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0746 - acc: 0.9768 - val_loss: 0.0970 - val_acc: 0.9705
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0630 - acc: 0.9803 - val_loss: 0.0893 - val_acc: 0.9740
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0558 - acc: 0.9820 - val_loss: 0.0992 - val_acc: 0.9720
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0482 - acc: 0.9846 - val_loss: 0.0873 - val_acc

#PReLU

keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)

Parametric Rectified Linear Unit.

It follows: f(x) = alpha * x for x < 0, f(x) = x for x >= 0, where alpha is a learned array with the same shape as x.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha_initializer: initializer function for the weights.
2. alpha_regularizer: regularizer for the weights.
3. alpha_constraint: constraint for the weights.
4. shared_axes: the axes along which to share learnable parameters for the activation function. For example, if the incoming feature maps are from a 2D convolution with output shape (batch, height, width, channels), and you wish to share parameters across space so that each filter only has one set of parameters, set shared_axes=[1, 2].

In [29]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['PReLU'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 0.2240 - acc: 0.9320 - val_loss: 0.1009 - val_acc: 0.9679
Epoch 2/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0836 - acc: 0.9737 - val_loss: 0.0932 - val_acc: 0.9717
Epoch 3/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0544 - acc: 0.9829 - val_loss: 0.0817 - val_acc: 0.9766
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0378 - acc: 0.9885 - val_loss: 0.0717 - val_acc: 0.9783
Epoch 5/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.0289 - acc: 0.9909 - val_loss: 0.0810 - val_acc: 0.9804
Epoch 6/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0224 - acc: 0.9930 - val_loss: 0.0837 - val_acc: 0.9807
Epoch 7/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.0178 - acc: 0.9944 - val_loss: 0.0861 - val_acc

#ELU

keras.layers.ELU(alpha=1.0)

Exponential Linear Unit.

It follows: f(x) =  alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha: scale for the negative factor.

In [30]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['ELU'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2840 - acc: 0.9131 - val_loss: 0.1398 - val_acc: 0.9544
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1188 - acc: 0.9628 - val_loss: 0.1016 - val_acc: 0.9683
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0812 - acc: 0.9750 - val_loss: 0.1080 - val_acc: 0.9684
Epoch 4/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0639 - acc: 0.9797 - val_loss: 0.0804 - val_acc: 0.9766
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0484 - acc: 0.9841 - val_loss: 0.0772 - val_acc: 0.9783
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0398 - acc: 0.9870 - val_loss: 0.1015 - val_acc: 0.9731
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0327 - acc: 0.9888 - val_loss: 0.0883 - val_acc

#ThresholdedReLU

keras.layers.ThresholdedReLU(theta=1.0)

Thresholded Rectified Linear Unit.

It follows: f(x) = x for x > theta, f(x) = 0 otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. theta: float >= 0. Threshold location of activation.

In [31]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['ThresholdedReLU'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 2.0977 - acc: 0.2059 - val_loss: 0.4568 - val_acc: 0.8796
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.2730 - acc: 0.9182 - val_loss: 0.1774 - val_acc: 0.9462
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1393 - acc: 0.9583 - val_loss: 0.1183 - val_acc: 0.9631
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0963 - acc: 0.9707 - val_loss: 0.0948 - val_acc: 0.9710
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0732 - acc: 0.9780 - val_loss: 0.0843 - val_acc: 0.9756
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0579 - acc: 0.9821 - val_loss: 0.0838 - val_acc: 0.9762
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0477 - acc: 0.9852 - val_loss: 0.0818 - val_acc

#Softmax

keras.layers.Softmax(axis=-1)

Softmax activation function.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. axis: Integer, axis along which the softmax normalization is applied.

In [32]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['Softmax'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 2.2696 - acc: 0.1896 - val_loss: 2.1972 - val_acc: 0.5439
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 2.0437 - acc: 0.7202 - val_loss: 1.8548 - val_acc: 0.7511
Epoch 3/10
60000/60000 [==============================] - 2s 33us/step - loss: 1.6432 - acc: 0.7527 - val_loss: 1.4281 - val_acc: 0.7519
Epoch 4/10
60000/60000 [==============================] - 2s 33us/step - loss: 1.2374 - acc: 0.7551 - val_loss: 1.0570 - val_acc: 0.7553
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9187 - acc: 0.7582 - val_loss: 0.8001 - val_acc: 0.7581
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.7188 - acc: 0.7606 - val_loss: 0.6545 - val_acc: 0.7621
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.6170 - acc: 0.7637 - val_loss: 0.5910 - val_acc

#ReLU

keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)

Rectified Linear Unit activation function.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = negative_slope * (x - threshold) otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. max_value: float >= 0. Maximum activation value.
2. negative_slope: float >= 0. Negative slope coefficient.
3. threshold: float. Threshold value for thresholded activation.

In [33]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
activations['ReLU'] = score[1]

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 0.2238 - acc: 0.9319 - val_loss: 0.1400 - val_acc: 0.9548
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0827 - acc: 0.9750 - val_loss: 0.0917 - val_acc: 0.9732
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0540 - acc: 0.9832 - val_loss: 0.0753 - val_acc: 0.9760
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0389 - acc: 0.9879 - val_loss: 0.0745 - val_acc: 0.9775
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0292 - acc: 0.9911 - val_loss: 0.0700 - val_acc: 0.9820
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0226 - acc: 0.9930 - val_loss: 0.0807 - val_acc: 0.9813
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0180 - acc: 0.9940 - val_loss: 0.1216 - val_acc

In [40]:
print(max(activations, key=activations.get), max(activations.values()))

ReLU 0.9842


ReLU has highest test accuracy of 98.42%

#Thank you for completing this notebook